### PR TITLE
[SILOptimizer] Added MandatoryCombiner to passes.

### PR DESCRIPTION
**Pull request by nate-chandler**
The mandatory combiner is added both to the Onone pass and also to the performance pass.

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please smoke test

File: lib/SILOptimizer/Mandatory/MandatoryCombiner.cpp [hunk 1]
```diff
@@ -143,20 +143,25 @@ struct MandatoryCombiner : SILFunctionTransform {
...
+        ii = next_or_default(ii, /*end=*/block.end(),
+                             /*defaultIter=*/block.begin());
         if (deleted) {
-          // The deletion succeeded.  The iterator has already been moved back
-          // a step so as to avoid invalidation.  Advancing it one step will
-          // not advance it to the current instruction since that has been
-          // deleted.  Instead, it will advance it to the next *remaining*
-          // instruction after the previous instruction as desired.
-          ii = next_or_default(ii, /*end*/block.end(), /*defaultIter*/block.begin());
+          // The current instruction has been deleted.  The iterator has already
```
Comment (gottesmm) [inline lib/SILOptimizer/Mandatory/MandatoryCombiner.cpp]: Can you move any changes that are not directly related to adding this to the pass pipeline into a different PR (or at least a different commit)?

File: lib/SILOptimizer/PassManager/PassPipeline.cpp [hunk 2]
```diff
@@ -658,6 +660,8 @@ SILPassPipelinePlan::getOnonePassPipeline(const SILOptions &Options) {
   P.startPipeline("Serialization");
   P.addSerializeSILPass();
 
+  P.addMandatoryCombiner();
```
Comment (gottesmm) [inline lib/SILOptimizer/PassManager/PassPipeline.cpp]: I think we want to add this before serialization.

File: lib/SILOptimizer/PassManager/PassPipeline.cpp [hunk 3]
```diff
@@ -393,6 +393,8 @@ static void addPerfEarlyModulePassPipeline(SILPassPipelinePlan &P) {
   // we do not spend time optimizing them.
   P.addDeadFunctionElimination();
 
+  P.addMandatoryCombiner();
```
Comment (gottesmm) [inline lib/SILOptimizer/PassManager/PassPipeline.cpp]: I am fine with this here, but this may make more sense in the SILOptPreparePassPipeline pipeline. Your thoughts?

File: lib/SILOptimizer/Mandatory/MandatoryCombiner.cpp (hunk 1 above)
Comment (nate-chandler) [reply lib/SILOptimizer/Mandatory/MandatoryCombiner.cpp]: Yup!  Will split into separate commit.

Comment (nate-chandler) [reply lib/SILOptimizer/Mandatory/MandatoryCombiner.cpp]: https://github.com/apple/swift/pull/26924/commits/e4cee7fdfaac4051b8b75f7f042f93d20f22bb1e

File: lib/SILOptimizer/PassManager/PassPipeline.cpp (hunk 2 above)
Comment (nate-chandler) [reply lib/SILOptimizer/PassManager/PassPipeline.cpp]: Fixed.

**Review by gottesmm — APPROVED** [review]
LGTM

File: lib/SILOptimizer/PassManager/PassPipeline.cpp (hunk 3 above)
Comment (nate-chandler) [reply lib/SILOptimizer/PassManager/PassPipeline.cpp]: It happening earlier seems only for the good.  Fixed.

**Comment by nate-chandler** [comment]
@swift-ci please test and merge

**Comment by gottesmm** [comment]
@swift-ci please test and merge

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please test and merge

**Comment by nate-chandler** [comment]
@swift-ci please test and merge
